File: 992/992.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/992.py

val MOD: Long = 987_898_789L

fun modPow(base: Long, exp: Long, modulus: Long): Long {
    var result = 1L
    var b = base % modulus
    var e = exp
    while (e > 0L) {
        if (e and 1L == 1L) result = result * b % modulus
        e = e shr 1
        b = b * b % modulus
    }
    return result
}

fun buildCombinatorics(limit: Int, modv: Long): Pair<LongArray, LongArray> {
    val fact = LongArray(limit + 1)
    fact[0] = 1L
    for (i in 1..limit) {
        fact[i] = fact[i - 1] * i.toLong() % modv
    }
    val invFact = LongArray(limit + 1)
    invFact[limit] = modPow(fact[limit], modv - 2L, modv)
    for (i in limit downTo 1) {
        invFact[i - 1] = invFact[i] * i.toLong() % modv
    }
    return Pair(fact, invFact)
}

class Comb(limit: Int, val modv: Long) {
    val fact: LongArray
    val invFact: LongArray

    init {
        val (f, inv) = buildCombinatorics(limit, modv)
        fact = f
        invFact = inv
    }

    fun call(n: Long, r: Long): Long {
        if (r < 0L || r > n) return 0L
        return fact[n.toInt()] * invFact[r.toInt()] % modv * invFact[(n - r).toInt()] % modv
    }
}

fun endpointCount(n: Int, k: Long, end: Int, comb: Comb, modv: Long): Long {
    if (n == 0) return 1L
    val right = LongArray(n)
    right[0] = k - (if (end == 0) 1L else 0L)
    if (n >= 2) {
        right[1] = 2L - (if (end == 1) 1L else 0L)
    }
    for (i in 2 until n) {
        right[i] = 1L + right[i - 2] - (if (end == i) 1L else 0L)
    }
    var ways = 1L
    for (v in 1 until n) {
        val outDegree = k + v.toLong() - (if (end == v) 1L else 0L)
        if (v < end) {
            ways = ways * comb.call(outDegree - 1L, right[v] - 1L) % modv
        } else if (v == end) {
            ways = ways * comb.call(outDegree, right[v]) % modv
        } else {
            ways = ways * comb.call(outDegree - 1L, right[v]) % modv
        }
    }
    return ways
}

fun journeyCount(n: Int, k: Long, comb: Comb, modv: Long): Long {
    var total = 0L
    for (end in 0..n) {
        total = (total + endpointCount(n, k, end, comb, modv)) % modv
    }
    return total
}

fun solve(): Long {
    val n = 500
    val ks = listOf(1L, 10L, 100L, 1000L, 10000L)
    val maxK = ks.max()!!
    val comb = Comb((maxK + n).toInt(), MOD)
    check(journeyCount(3, 2L, comb, MOD) == 17L)
    check(journeyCount(6, 1L, comb, MOD) == 1320L)
    check(journeyCount(6, 5L, comb, MOD) == 16_793_280L)
    var answer = 0L
    for (k in ks) {
        answer = (answer + journeyCount(n, k, comb, MOD)) % MOD
    }
    return answer
}

fun main() {
    println(solve())
}

File: 992/992.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/992.py

const MOD: u64 = 987_898_789;

fn build_combinatorics(limit: usize, modv: u64) -> (Vec<u64>, Vec<u64>) {
    let mut fact = vec![1u64; limit + 1];
    for i in 1..=limit {
        fact[i] = fact[i - 1] * (i as u64) % modv;
    }
    let mut inv_fact = vec![1u64; limit + 1];
    inv_fact[limit] = mod_pow(fact[limit], modv - 2, modv);
    for i in (1..=limit).rev() {
        inv_fact[i - 1] = inv_fact[i] * (i as u64) % modv;
    }
    (fact, inv_fact)
}

fn mod_pow(mut base: u64, mut exp: u64, modulus: u64) -> u64 {
    let mut result = 1u64;
    base %= modulus;
    while exp > 0 {
        if exp & 1 == 1 {
            result = result * base % modulus;
        }
        exp >>= 1;
        base = base * base % modulus;
    }
    result
}

struct Comb {
    modv: u64,
    fact: Vec<u64>,
    inv_fact: Vec<u64>,
}

impl Comb {
    fn new(limit: usize, modv: u64) -> Self {
        let (fact, inv_fact) = build_combinatorics(limit, modv);
        Comb { modv, fact, inv_fact }
    }

    fn call(&self, n: i64, r: i64) -> u64 {
        if r < 0 || r > n {
            return 0;
        }
        let n = n as usize;
        let r = r as usize;
        self.fact[n] * self.inv_fact[r] % self.modv * self.inv_fact[n - r] % self.modv
    }
}

fn endpoint_count(n: usize, k: i64, end: usize, comb: &Comb, modv: u64) -> u64 {
    if n == 0 {
        return 1;
    }
    let mut right = vec![0i64; n];
    right[0] = k - (if end == 0 { 1 } else { 0 });
    if n >= 2 {
        right[1] = 2 - (if end == 1 { 1 } else { 0 });
    }
    for i in 2..n {
        right[i] = 1 + right[i - 2] - (if end == i { 1 } else { 0 });
    }
    let mut ways = 1u64;
    for v in 1..n {
        let out_degree = k + v as i64 - (if end == v { 1 } else { 0 });
        if v < end {
            ways = ways * comb.call(out_degree - 1, right[v] - 1) % modv;
        } else if v == end {
            ways = ways * comb.call(out_degree, right[v]) % modv;
        } else {
            ways = ways * comb.call(out_degree - 1, right[v]) % modv;
        }
    }
    ways
}

fn journey_count(n: usize, k: i64, comb: &Comb, modv: u64) -> u64 {
    let mut total = 0u64;
    for end in 0..=(n as usize) {
        total = (total + endpoint_count(n, k, end, comb, modv)) % modv;
    }
    total
}

fn solve() -> u64 {
    let n: usize = 500;
    let ks: Vec<i64> = vec![1, 10, 100, 1000, 10000];
    let max_k = *ks.iter().max().unwrap();
    let comb = Comb::new((max_k as usize) + n, MOD);
    assert_eq!(journey_count(3, 2, &comb, MOD), 17);
    assert_eq!(journey_count(6, 1, &comb, MOD), 1320);
    assert_eq!(journey_count(6, 5, &comb, MOD), 16_793_280);
    let mut answer = 0u64;
    for k in &ks {
        answer = (answer + journey_count(n, *k, &comb, MOD)) % MOD;
    }
    answer
}

fn main() {
    println!("{}", solve());
}
